### PR TITLE
Autotools for iOS

### PR DESCRIPTION
**Pull request by pretyman**
* cmake/modules/hunter_autotools_project.cmake
Moved common architecture logic up, specific logic down

*  scripts/autotools-merge-lipo.cmake.in
Broke the horrible big `foreach` loop into functions to become somewhat more readable
Removed *.la files from libtool 
Added a descriptive header to the generated header files and descriptive error message

### CODE REVIEW

**Comment by pretyman** [comment]
So I just discovered from reading `lipo` man page that I can `lipo` executable binaries together, not only static libraries. Maybe I can `lipo` anything generated in the `${architecture}/bin` directory into a single binary instead of appending the architecture. I was trouble determining what would be binary but I think I can use the return code of `lipo -info` for that. I can do the change in a later refactor, it seems sensible.

File: scripts/autotools-merge-lipo.cmake.in [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+# Remove the library files
+foreach(x ${library_files})
+  list(REMOVE_ITEM all_except_library_files ${x})
+endforeach()
+
+
+# The preprocessor macros below are from
+# http://sourceforge.net/p/predef/wiki/Architectures/
+# except armv7, which was taken from:
+# clang -arch armv7 -dD -E config.h
```
Comment (ruslo) [inline scripts/autotools-merge-lipo.cmake.in]: Cool, was not thinking about '-arch'.

Improved script: https://github.com/ruslo/hunter/commit/9e0494913b652ef66b842e3e0609c48d2ba6cb3a

Updated:
 * https://github.com/ruslo/hunter/commit/3a47834642ac5f3c9765a3982ee955c0ba8e48ef
 * https://github.com/ruslo/hunter/commit/83572273937bf759449cfa8430358503a6bbf7d0
 * https://github.com/ruslo/hunter/commit/f4283306fe15d82297fd28f3a101c19f61847046
 * https://github.com/ruslo/hunter/commit/24aab1606b76b4e57e68b2e008aa098dfe8d14ce

**Comment by ruslo** [comment]
Merged to `ios-dev` branch for now with some improvements:
* https://github.com/ruslo/hunter/commit/56e295821219b07e5c7f0192c7ea137311a317c3
* https://github.com/ruslo/hunter/commit/81fd1a98e862d861434de88e40a4b2a4a39f2479?diff=split
* https://github.com/ruslo/hunter/commit/a3c491a226241f4b938df4a6a4bf9b0cc3dd148e
* https://github.com/ruslo/hunter/commit/11e2bedcf0683e2aa0bf123ea03659114a856d80?diff=split
* https://github.com/ruslo/hunter/commit/3aab6f8470f5c2585aa9b13ec6aa7c01da636292
* https://github.com/ruslo/hunter/commit/a792a40b290cbccc9173cfa2859140aa281da6b5

want to test it a little bit more

**Comment by ruslo** [comment]
> want to test it a little bit more

Add test: https://github.com/ruslo/hunter/commit/5c126f9c787d61bac752797c1d116afdc27cca54

Fix (header `File auto generated by Hunter...` is lost): https://github.com/ruslo/hunter/commit/0cdcb09a2fef80e13a630c6a8b6e4a8d9d879f4a

**Comment by ruslo** [comment]
Fix for case when header for one arch is missing: https://github.com/ruslo/hunter/commit/27961a83f01e38442d45b937f2e9e40663892f85

**Comment by pretyman** [comment]
> We should use HUNTER_PACKAGE_BUILD_DIR 

Ok, the first version I did was with `HUNTER_PACKAGE_BUILD_DIR`, however I wanted to have a way to investigate, in a multi-arch built, if a build created files with `somefile-arch`. Doesn't  `HUNTER_PACKAGE_BUILD_DIR` get deleted upon completion? If so, it becomes hard to understand and be sure if these kind of files were not created. We could also have a first target to delete the multi-arch root as a way to guarantee a clean condition in the edge case you brought up.

**Comment by ruslo** [comment]
> however I wanted to have a way to investigate, in a multi-arch built, if a build created files with the 

I'm not sure I understand this, can you clarify? Why you can't do `if(EXISTS ".../somefile-arch")`?

> We could also have a first target to delete the multi-arch root as a way to guarantee a clean condition in the edge case you brought up

Yes, this will work too I guess.

> it is easy to understand what differed from one build to another

How about checking at the end that `<root>/multi-arch` is a directory with no files. This is an easy work for `file(GLOB_RECURSE ...)` command.

**Comment by pretyman** [comment]
Thanks for all your fixes, I learnt a lot with them. 

> I'm not sure I understand this, can you clarify?

Ok, sorry for not being clear, let describe the use case: this is at Hunter package creation/update time, when we need to introduce new packages, without ever having done a multi-arch built on them before (or without ever cross compiled to different arch, like `arm`), hence we don't know yet the very exact outcome of the build, and would be nice to guarantee that no file names dependent on the architecture were made.

This is because of the simple algorithm used to merge to builds, which picks one result of a build to create a list of the available files to be merged, however if the filenames differ in other architectures' build, it would be nice to have a way to easily understand that this case has happened.

> How about checking at the end that <root>/multi-arch is a directory with no files. This is an easy work for file(GLOB_RECURSE ...) command. 

Yeah, that is a good idea. If we `message(FATAL_ERROR ...)` from there, I believe the directory won't be deleted, making it possible to investigate the differences, right?

**Comment by ruslo** [comment]
> let describe the use case

Still don't get it. Why we need to use final destination HUNTER_PACKAGE_INSTALL_DIR instead of temporary one HUNTER_PACKAGE_BUILD_DIR (note that for the uncacheable packages HUNTER_PACKAGE_INSTALL_DIR is equal to HUNTER_INSTALL_DIR).

Okay, so what about:
* We use HUNTER_PACKAGE_BUILD_DIR
* HUNTER_PACKAGE_BUILD_DIR cleaned before and after (successful) package creation
* install each arch to HUNTER_PACKAGE_BUILD_DIR/${arch}
* lipo/merge files to HUNTER_PACKAGE_BUILD_DIR/multi-arch
* check there are no files left in HUNTER_PACKAGE_BUILD_DIR/${arch}, on error this directory will not be deleted
* install files to the final HUNTER_PACKAGE_INSTALL_DIR directory by simply copying everything from HUNTER_PACKAGE_BUILD_DIR/multi-arch

Does it sounds good? Do you see any flaws in this design?

**Comment by pretyman** [comment]
Sounds perfect, your proposal is great. I can implement it in a near future. Just let me know when you are done with the work in the `ios-dev` branch, and I will continue from your branch. I just don't want to handle any conflicts when I can focus other things on my side.

Thanks

**Comment by ruslo** [comment]
> I can implement it in a near future. Just let me know when you are done with the work in the ios-dev branch

Cool, I'm implementing it now.

**Comment by ruslo** [comment]
> Cool, I'm implementing it now.

see https://github.com/ruslo/hunter/commit/39dd60475448d8f4adc7c99ff7d67b570d22e5f0

**Comment by ruslo** [comment]
> checking at the end that <root>/multi-arch is a directory with no files

implemented: https://github.com/ruslo/hunter/commit/e58b72cd26d16cfe29e85dc4a77e6da4dd004bd7

**Comment by pretyman** [comment]
Круто! Очень хорошо! Thank you so much!

**Comment by ruslo** [comment]
> So I just discovered from reading lipo man page that I can lipo executable binaries together, not only static libraries

applied: https://github.com/ruslo/hunter/commit/0d30d1c8b5c8b16d964e25c7b0b91125e54d0671

**Pull request merged**